pkg/reconciler/tenancy/workspacemounts/workspacemounts_controller_workspaces.go

```diff
@@ -0,0 +1,208 @@
+		queue:    queue,
+		gvrQueue: gvkQueue,
+
+		dynamicClusterClient:                    dynamicClusterClient,
+		discoveringDynamicSharedInformerFactory: discoveringDynamicSharedInformerFactory,
+
+		workspaceIndexer: workspaceInformer.Informer().GetIndexer(),
+		workspaceLister:  workspaceInformer.Lister(),
+
+		commit: committer.NewCommitter[*tenancyv1alpha1.Workspace, tenancyv1alpha1client.WorkspaceInterface, *tenancyv1alpha1.WorkspaceSpec, *tenancyv1alpha1.WorkspaceStatus](kcpClusterClient.TenancyV1alpha1().Workspaces()),
+	}
+
+	_, _ = workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+	})
+
+	c.discoveringDynamicSharedInformerFactory.AddEventHandler(informer.GVREventHandlerFuncs{
+		AddFunc:    func(gvr schema.GroupVersionResource, obj interface{}) { c.enqueueForResource(logger, gvr, obj) },
+		UpdateFunc: func(gvr schema.GroupVersionResource, _, obj interface{}) { c.enqueueForResource(logger, gvr, obj) },
+		DeleteFunc: nil, // Nothing to do.
+	})
+
+	return c, nil
+}
+
+type workspaceResource = committer.Resource[*tenancyv1alpha1.WorkspaceSpec, *tenancyv1alpha1.WorkspaceStatus]
+
+// Controller watches Workspaces and dynamically discovered mount resources and reconciles them so
+// workspace has right annotations.
+type Controller struct {
+	// queue is the work-queue used by the controller for workspaces
+	queue workqueue.RateLimitingInterface
+	// gvrQueue is the work-queue used by the controller for dynamically discovered mount resources.
+	gvrQueue workqueue.RateLimitingInterface
+
+	dynamicClusterClient                    kcpdynamic.ClusterInterface
+	discoveringDynamicSharedInformerFactory *informer.DiscoveringDynamicSharedInformerFactory
+
+	workspaceIndexer cache.Indexer
+	workspaceLister  tenancyv1alpha1listers.WorkspaceClusterLister
+
+	// commit creates a patch and submits it, if needed.
+	commit func(ctx context.Context, new, old *workspaceResource) error
+}
+
+// enqueue adds the object to the work queue, used for workspaces.
+func (c *Controller) enqueue(obj interface{}) {
+	key, err := kcpcache.MetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(2).Info("queueing Workspace")
+	c.queue.Add(key)
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+	defer c.gvrQueue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+		go wait.Until(func() { c.startGVKWorker(ctx) }, time.Second, ctx.Done())
```

Comment:
as written above, get rid of this second controller. Everything done in there is in-memory and does not block. Do it directly from the dynamic shared event handler queuing into the main queue.